### PR TITLE
Use New Package Structure From Original Shine And Fix Bug in CUDA Codegen

### PR DESCRIPTION
**Pull request by bastian-koepcke**
Before, the CUDA code generator would indirectly recursively call a code generation function again.
Due to a missing overload this lead to a dispatch to an OpenCL code generation function which does not know how to deal with the CUDA primitives.
Simply overloading the function should fix it.

Simultaneously, this merges the newest changes of the master branch of the original shine repository. 
* This is changing some package names in order to fix problems that occur in newer IntelliJ versions due to name overloading.
* This is changing how submodules are organized.
* It also adds a scala style config to aid in sticking to the code style of the original repository (but nothing is enforced).

What is *not* included in this PR:
* A fix for the test `amount_vector_equals_1_secondversion`.
* A fix for the test `nbody versions produce same results` which is failing in this fork but not in the original repository (there are more tests with similar errors).

I believe the following things are important for you to look at:
* The `yacx` commit that is going to be used, maybe you want to use `HEAD` in `master` instead.
* Correctness of changes to `build.sbt`.

The rest of the changes you can most likely ignore because they are straightforward changes from the original repository.

### CODE REVIEW

**Review by bastian-koepcke — COMMENTED** [review]
I am not sure about the used commit in `yacx`. The bug was shown to me specific to a commit, but not a branch. It may be better to use the HEAD of master of `yacx`, but when I tried the Java Files were not found during compilation.

**Review by LukasSiefke — APPROVED** [review]

File: build.sbt [hunk 1]
```diff
@@ -42,6 +42,7 @@ lazy val shine = (project in file("."))
   )
 
 lazy val executor   = (project in file("lib/executor"))
```
Comment (zeratax) [inline build.sbt:44]: nebenbei, aber wäre halt wirklich schön wenn das eine CLexecutor und das andere CUexecutor wäre

File: src/main/scala/shine/OpenCL/CodeGeneration/CodeGenerator.scala [hunk 2]
```diff
@@ -218,6 +219,7 @@ class CodeGenerator(override val decls: CCodeGenerator.Declarations,
 
   override def genNat(n: Nat, env: Environment, cont:Expr => Stmt): Stmt = n match {
     case of: BuiltInFunctionCall => cont(C.AST.Literal(of.toString))
+    case of: BuiltInAttribute => cont(C.AST.Literal(of.toString))
```
Comment (zeratax) [inline src/main/scala/shine/OpenCL/CodeGeneration/CodeGenerator.scala:222]: etwas komisch das hier zu haben, aber gut es funktioniert ja

File: src/main/scala/shine/OpenCL/AdjustArraySizesForAllocations.scala [hunk 3]
```diff
@@ -34,6 +35,9 @@ object AdjustArraySizesForAllocations {
       case mG@MapGlobal(dim) => visitAndGatherInformation(mG.f, BasicInfo(Global, dim) :: parallInfo)
       case mWG@MapWorkGroup(dim) => visitAndGatherInformation(mWG.f, BasicInfo(WorkGroup, dim) :: parallInfo)
       case mL@MapLocal(dim) => visitAndGatherInformation(mL.f, BasicInfo(Local, dim) :: parallInfo)
+      case mG@MapThreads(dim) => visitAndGatherInformation(mG.f, BasicInfo(Global, dim) :: parallInfo)
+      case mWG@MapGrid(dim) => visitAndGatherInformation(mWG.f, BasicInfo(WorkGroup, dim) :: parallInfo)
+      case mL@MapBlock(dim) => visitAndGatherInformation(mL.f, BasicInfo(Local, dim) :: parallInfo)
```
Comment (zeratax) [inline src/main/scala/shine/OpenCL/AdjustArraySizesForAllocations.scala:40]: genauso das hier

**Review by zeratax — APPROVED** [review]

**Comment by zeratax** [comment]
wir können das yacx submodule jetzt auch auf den master branch wechseln da https://github.com/ZerataX/yacx/pull/138 gemerged ist

**Pull request merged**